subsys: dfu: default progressive erase on Nordic

pabigot opened the pull request:
The instructions for samples/usb/dfu fail on Nordic platforms if the erase is not progressive.  Default to enable that on Nordic.

This allows `samples/usb/dfu` to work as described on Nordic platforms.

ulfalizer commented on subsys/dfu/Kconfig:
```diff
@@ -53,6 +53,7 @@ config IMG_BLOCK_BUF_SIZE
 config 	IMG_ERASE_PROGRESSIVELY
```
Could maybe put a single space here at the same time.

nvlsianpu commented:
@pdunaj Are you fine with this patch?

pdunaj commented:
Hi @nvlsianpu , we don't use this API so it should not affect us.

nvlsianpu commented:
@de-nordic Any obstacles from the mcumgr side?

de-nordic commented:
> 
> 
> @de-nordic Any obstacles from the mcumgr side?
I do not think so.

de-nordic commented:
> 
> 
> The instructions for samples/usb/dfu fail on Nordic platforms if the erase is not progressive. Default to enable that on Nordic.
> 
> This allows `samples/usb/dfu` to work as described on Nordic platforms.

Why does it fail? It timeouts?

pabigot commented:
> Why does it fail? It timeouts?

Probably, though the diagnostic is just "fails".  Fortunately the slack `#usb` message is still visible:
```
tirzah[27]$ dfu-util --alt 1 --download build/zephyr/zephyr.signed.bin
dfu-util 0.9
Copyright 2005-2009 Weston Schmidt, Harald Welte and OpenMoko Inc.
Copyright 2010-2016 Tormod Volden and Stefan Schmidt
This program is Free Software and has ABSOLUTELY NO WARRANTY
Please report bugs to http://sourceforge.net/p/dfu-util/tickets/
dfu-util: Invalid DFU suffix signature
dfu-util: A valid DFU suffix will be required in a future dfu-util release!!!
Opening DFU capable USB device...
ID 2fe3:0005
Run-time device DFU version 0110
Claiming USB DFU Runtime Interface...
Determining device status: state = appIDLE, status = 0
Device really in Runtime Mode, send DFU detach request...
Resetting USB...
Opening DFU USB Device...
Claiming USB DFU Interface...
Setting Alternate Setting #1 ...
Determining device status: state = dfuIDLE, status = 0
dfuIDLE, continuing
DFU mode device DFU version 0110
Device returned transfer size 128
Copying data from PC to DFU device
Download    [                         ]   0%            0 bytesdfu-util: Error during download get_status
tirzah[28]$
```

de-nordic commented:
> 
> 
> > Why does it fail? It timeouts?
> 
> Probably, though the diagnostic is just "fails". Fortunately the slack `#usb` message is still visible:
> 
> ```
> tirzah[27]$ dfu-util --alt 1 --download build/zephyr/zephyr.signed.bin
> dfu-util 0.9
> Copyright 2005-2009 Weston Schmidt, Harald Welte and OpenMoko Inc.
> Copyright 2010-2016 Tormod Volden and Stefan Schmidt
> This program is Free Software and has ABSOLUTELY NO WARRANTY
> Please report bugs to http://sourceforge.net/p/dfu-util/tickets/
> dfu-util: Invalid DFU suffix signature
> dfu-util: A valid DFU suffix will be required in a future dfu-util release!!!
> Opening DFU capable USB device...
> ID 2fe3:0005
> Run-time device DFU version 0110
> Claiming USB DFU Runtime Interface...
> Determining device status: state = appIDLE, status = 0
> Device really in Runtime Mode, send DFU detach request...
> Resetting USB...
> Opening DFU USB Device...
> Claiming USB DFU Interface...
> Setting Alternate Setting #1 ...
> Determining device status: state = dfuIDLE, status = 0
> dfuIDLE, continuing
> DFU mode device DFU version 0110
> Device returned transfer size 128
> Copying data from PC to DFU device
> Download    [                         ]   0%            0 bytesdfu-util: Error during download get_status
> tirzah[28]$
> ```

--verbose does not give more info?

pabigot commented:
> --verbose does not give more info?

You're welcome to try it and search for alternative approaches if you don't like this solution.  I was told this enabling progressive erase was required on Nordic, and since that wasn't documented or implemented anywhere obvious thought it was helpful to make it the default.

nixward commented:
> > --verbose does not give more info?
> 
> I was told this enabling progressive erase was required on Nordic, and since that wasn't documented or implemented anywhere obvious thought it was helpful to make it the default.

@pabigot did they say this requirement is specific for USB DFU?  From experience with Zephyr 1.14.x BLE DFU progressive erase hasn't been a requirement.

pabigot commented:
@nixward If you have slack access see https://zephyrproject.slack.com/archives/CDUAL494Y/p1573846077137200 but that will become inaccessable in a few weeks.

pabigot commented:
I think I understand the concern: progressive erase is apparently not required when using DFU over BLE, only over USB.

If there's a real concern that this could cause a problem with BLE updates, please investigate.  If it is, we can address it by documentation, though it'll probably continue to bite people (because who reads documentation?).

de-nordic commented:
> 
> 
> I think I understand the concern: progressive erase is apparently not required when using DFU over BLE, only over USB.
> 
> If there's a real concern that this could cause a problem with BLE updates, please investigate. If it is, we can address it by documentation, though it'll probably continue to bite people (because who reads documentation?).

Progressive erase is used, as far as I can tell, with Ble and serial update with mcumgr lib.
Miguel (mlaz from Juul i think) is working on fixing broken mcumgrlib that uses progressive erase, maybe progressive erase should be set as default for flashing updates. Transfer without progressive erase takes longer, maybe the same thing happens here.

de-nordic reviewed (approved):
Checked with current version of mcumgrlib. No negative impact. There is no reason to hold it back.

nixward reviewed (changes requested):
I think it would be better if Kconfig expressed the absolute requirement that nRF USB DFU has on the IMG_ERASE_PROGRESSIVELY symbol rather than it just becoming the default for all nRF DFU.  If the BLE DFU has no requirement for IMG_ERASE_PROGRESSIVELY then it should stay the same with a slightly smaller image size as result.  The user can then opt in they choose.

It seems the IMG_ERASE_PROGRESSIVELY symbol is a requirement for USB DFU and a nice, possibly more performant option for other DFU transports.

nixward commented:
For reference @nvlsianpu PR for IMG_ERASE_PROGRESSIVELY
https://github.com/zephyrproject-rtos/zephyr/commit/dffcfd8c79552c7ca48f0ec4f32f06b183d5cd87#diff-c7dc166e32833e098bb451e68ecea3fd

pabigot commented:
@nvlsianpu @jfischer-phytec-iot @de-nordic do you concur with making this additionally depend on `USB_DFU_CLASS`?  The space savings for non-USB DFU would be 421 bytes of flash by my measurement.

pabigot committed:
usb: dfu: default progressive erase on Nordic

The instructions for samples/usb/dfu fail on Nordic platforms if the
erase is not progressive.  Default to enable that on Nordic.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

nvlsianpu commented:
@pabigot - sounds good

pabigot commented:
@nixward Updated to add USB DFU dependency.

nixward commented:
@pabigot I'm not an expert in Kconfig but can this be a select instead of depends?  The new logic still implies the user can opt out of this option when SOC_FLASH_NRF && USB_DFU_CLASS when it seems they can't.

pabigot commented:
> @pabigot I'm not an expert in Kconfig but can this be a select instead of depends? The new logic still implies the user can opt out of this option when SOC_FLASH_NRF && USB_DFU_CLASS when it seems they can't.

@nixward I don't understand your question.  There is no change to select/depend usage here, and I don't intend to make one.  The sole intent of this change is to ensure that a system that requires progressive erase to function enables it by default.  If a user chooses to override that default for some reason I don't see why we should prevent that.

nvlsianpu commented:
@nixward For some cases user (even nRF user) might want disable progressive erase (especially while code partition is small, so erase time is below 5 seconds).

nixward commented:
> @nixward For some cases user (even nRF user) might want disable progressive erase (especially while code partition is small, so erase time is below 5 seconds).

@nvlsianpu agreed.  I have no problem with the opting in or out of progressive erase when the DFU will work either way.

nixward commented:
> The sole intent of this change is to ensure that a system that requires progressive erase to function enables it by default. If a user chooses to override that default for some reason I don't see why we should prevent that.

@pabigot the issue I see is we should be protecting the end user from selecting the following config if we know the DFU process will be broken:
```
CONFIG_SOC_FLASH_NRF=y
CONFIG_USB_DFU_CLASS=y
CONFIG_IMG_ERASE_PROGRESSIVELY=n
```

So `CONFIG_IMG_ERASE_PROGRESSIVELY=y` shouldn't just be the default if 
`CONFIG_SOC_FLASH_NRF=y` and
`CONFIG_USB_DFU_CLASS=y`
it should be the only choice.

If the user chooses to modify Zephyr to override this it is their fault but it they think `CONFIG_IMG_ERASE_PROGRESSIVELY=y` is optional and choose `CONFIG_IMG_ERASE_PROGRESSIVELY=n` they'll see it as Zephyr's fault.

So from my understanding of Kconfig if your code change uses 'select' instead of 'default' then we will protect the Zephyr user from a potential DFU breaking config.

pabigot commented:
> > The sole intent of this change is to ensure that a system that requires progressive erase to function enables it by default. If a user chooses to override that default for some reason I don't see why we should prevent that.
> 
> the issue I see is we should be protecting the end user from selecting the following config if we know the DFU process will be broken:
> 
> ```
> CONFIG_SOC_FLASH_NRF=y
> CONFIG_USB_DFU_CLASS=y
> CONFIG_IMG_ERASE_PROGRESSIVELY=n
> ```
> 
> So `CONFIG_IMG_ERASE_PROGRESSIVELY=y` shouldn't just be the default if
> `CONFIG_SOC_FLASH_NRF=y` and
> `CONFIG_USB_DFU_CLASS=y`
> it should be the only choice.

But in https://github.com/zephyrproject-rtos/zephyr/pull/20737#issuecomment-555931895 you seem to acknowledge that there are situations where disabling progressive erase with USB DFU is a reasonable thing to do.

So no, I do not think switching to `select` is correct here.

nixward commented:
> 
> But in [#20737 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/20737#issuecomment-555931895) you seem to acknowledge that there are situations where disabling progressive erase with USB DFU is a reasonable thing to do.
> 
> So no, I do not think switching to `select` is correct here.

Yes I do think progressive erase should be an option for nRF BLE DFU (but not default).
I also think progressive erase should be the only option (enforced) for the nRF USB DFU.  The issue is the PR as it stands only makes progressive scan the default for the nRF USB DFU.  The user can still opt out (and wonder why the USB DFU now doesn't work) with `CONFIG_IMG_ERASE_PROGRESSIVELY=n`.

I think this can be achieved with the following change:
`select y if SOC_FLASH_NRF && USB_DFU_CLASS`

jfischer-phytec-iot commented:
>  The user can still opt out (and wonder why the USB DFU now doesn't work) with `CONFIG_IMG_ERASE_PROGRESSIVELY=n`

Why should not the user be allowed to do it? Maybe one want to try it out for some reason. Power To The People!

pabigot commented:
Added dev-review since we are unable to reach agreement on how to proceed.  To discuss tomorrow.

nixward commented:
> > The user can still opt out (and wonder why the USB DFU now doesn't work) with `CONFIG_IMG_ERASE_PROGRESSIVELY=n`
> 
> Why should not the user be allowed to do it? Maybe one want to try it out for some reason. Power To The People!

@jfischer-phytec-iot I don't disagree here, they have the source they can always remove this Kconfig line entirely and try it but I don't believe end users of Zephyr (myself included) that are possibly trying to reduce their image size need to trip into this hole.  This quite a potentially painful hole as well.  If a product is relying on USB DFU and a developer finds a menconf option to tweak the image size without checking that the USB DFU process still works, they will soon find that after the next DFU the DFU process is broken.

There is other logic used in Kconfig to stop users falling into holes, the symbol CONFIG_SOC_NRF52832_ALLOW_SPIM_DESPITE_PAN_58 for example.

At a minimum we should open an issue that nRF USB DFU doesn't work with CONFIG_IMG_ERASE_PROGRESSIVELY=n.  If it can be resolved then we can close the bug and remove the restriction.

pabigot commented:
> I think this can be achieved with the following change:
> `select y if SOC_FLASH_NRF && USB_DFU_CLASS`

That doesn't work, but I've replaced the original with an alternative that does force use of progressive erase on Nordic when using USB DFU.

I've re-requested reviews from everybody that had approved the old approach since this is not a drop-in replacement.

de-nordic reviewed (approved):
I think it is fine.

nixward commented:
Looks good @pabigot, thanks!  I tested with subsys/usb/dfu and subsys/mgmt/mcumgr/smp_svr samples with the nrf52840_10056 board and the Kconfig logic you have coded works.
